exwm-workspace.el

```diff
@@ -252,26 +281,106 @@ The optional FORCE option is for internal use only."
+    (xcb:+request exwm--connection
+        (make-instance 'xcb:ewmh:set-_NET_VIRTUAL_ROOTS
+                       :window exwm--root
+                       :data (vconcat (mapcar
+                                       (lambda (i)
+                                         (frame-parameter i 'exwm-window-id))
+                                       exwm-workspace--list))))
+    ;; Set _NET_NUMBER_OF_DESKTOPS
+    (xcb:+request exwm--connection
+        (make-instance 'xcb:ewmh:set-_NET_NUMBER_OF_DESKTOPS
+                       :window exwm--root
+                       :data num-workspaces))
+    ;; Set _NET_DESKTOP_VIEWPORT
+    (xcb:+request exwm--connection
+        (make-instance 'xcb:ewmh:set-_NET_DESKTOP_VIEWPORT
+                       :window exwm--root
+                       :data (make-vector (* 2 num-workspaces) 0)))
+
+    ;; Set _NET_WORKAREA (with minibuffer and bottom mode-line excluded)
+    (let* ((workareas
+            (vector 0 0 (x-display-pixel-width) (x-display-pixel-height)))
+           (workareas (mapconcat (lambda (i) workareas)
+                                 (make-list num-workspaces 0) [])))
+      (xcb:+request exwm--connection
+          (make-instance 'xcb:ewmh:set-_NET_WORKAREA
+                         :window exwm--root
+                         :data workareas)))
+    (xcb:flush exwm--connection)))
+
+(defun exwm-workspace--init ()
+  "Initialize workspace module."
+  ;; Prevent unexpected exit
+  (setq confirm-kill-emacs
+        (lambda (prompt)
+          (pcase (length exwm--id-buffer-alist)
+            (0 (y-or-n-p prompt))
+            (x (yes-or-no-p (format "[EXWM] %d window%s currently alive. %s"
+                                    x (if (= x 1) "" "s") prompt))))))
+  ;; Initialize workspaces
+  (let ((initial-workspaces (frame-list)))
+    (when (< 1 (length initial-workspaces))
+      ;; Emacs client creates an extra (but unusable) frame
+      (dolist (i initial-workspaces)
+        (unless (frame-parameter i 'window-id)
+          (setq initial-workspaces (delq i initial-workspaces))))
+      ;; Prevent user from deleting the first frame by accident
+      (set-frame-parameter (car initial-workspaces) 'client nil))
+    ;; Configure workspaces
+    (dolist (i initial-workspaces)
+      (exwm-workspace--add-frame-as-workspace i))
+    (select-frame-set-input-focus (car initial-workspaces)))
   ;; Handle unexpected frame switch
   (add-hook 'focus-in-hook 'exwm-workspace--on-focus-in)
-  ;; Set _NET_VIRTUAL_ROOTS
-  (xcb:+request exwm--connection
-      (make-instance 'xcb:ewmh:set-_NET_VIRTUAL_ROOTS
-                     :window exwm--root
-                     :data (vconcat (mapcar
-                                     (lambda (i)
-                                       (frame-parameter i 'exwm-window-id))
-                                     exwm-workspace--list))))
+  ;; Make new frames create new workspaces.
+  (setq window-system-default-frame-alist '((x . ((visibility . nil)))))
```

Comment:
I think you have an `rassq` for an `assq`, and the `setq` should be a push, but I might be wrong.

``` elisp
  (if (assq 'x window-system-default-frame-alist)
      (push '(visibility . nil) (cdr (assq 'x window-system-default-frame-alist)))
    (push '(x . ((visibility))) window-system-default-frame-alist))
```

maybe.